src/runtime_src/core/common/usage_metrics.cpp

```diff
@@ -0,0 +1,522 @@
+  }
+}
+
+static bpt::ptree
+get_bos_ptree(const bo_metrics& bo_met)
+{
+  bpt::ptree bo_tree;
+
+  bo_tree.add("total_count", bo_met.total_count);
+  bo_tree.add("size", std::to_string(bo_met.total_size_in_bytes) + " bytes");
+  bo_tree.add("avg_size", std::to_string(bo_met.total_size_in_bytes / bo_met.total_count) + " bytes");
+  bo_tree.add("peak_size", std::to_string(bo_met.peak_size_in_bytes) + " bytes");
+  bo_tree.add("bytes_synced_to_device", std::to_string(bo_met.bytes_synced_to_device) + " bytes");
+  bo_tree.add("bytes_synced_from_device", std::to_string(bo_met.bytes_synced_from_device) + " bytes");
+
+  return bo_tree;
+}
+
+static bpt::ptree
+get_kernels_ptree(const std::vector<kernel_metrics>& kernels_vec)
+{
+  bpt::ptree kernel_array;
+
+  for (const auto& kernel : kernels_vec) {
+    bpt::ptree kernel_tree;
+
+    kernel_tree.put("name", kernel.handle);
+    kernel_tree.put("num_of_args", kernel.num_args);
+    kernel_tree.put("num_total_runs", std::to_string(kernel.total_runs));
+
+    if (kernel.total_runs > 0) {
+      auto avg_run_time = (kernel.total_time.count()) / kernel.total_runs;
+      kernel_tree.put("avg_run_time", std::to_string(avg_run_time) + " us");
+    }
+    kernel_array.push_back(std::make_pair("", kernel_tree));
+  }
+
+  return kernel_array; 
+}
+
+static bpt::ptree
+get_hw_ctx_ptree(const std::vector<hw_ctx_metrics>& hw_ctx_vec)
+{
+  bpt::ptree hw_ctx_array;
+
+  uint32_t ctx_count = 0;
+  for (const auto& ctx : hw_ctx_vec) {
+    bpt::ptree hw_ctx;
+    hw_ctx.put("id", std::to_string(ctx_count));
+    hw_ctx.put("xclbin_uuid", ctx.xclbin_uuid.to_string());
+
+    // add buffer info
+    hw_ctx.add_child("bos", get_bos_ptree(ctx.bos_met));
+
+    // add kernel info
+    hw_ctx.add_child("kernels", get_kernels_ptree(ctx.kernel_metrics_vec));
+
+    hw_ctx_array.push_back(std::make_pair("", hw_ctx));
+    ctx_count++;
+  }
+
+  return hw_ctx_array;
+}
+
+static void
+print_usage_metrics()
+{
+  bpt::ptree thread_array;
+
+  uint32_t t_count = 0;
+  // iterate over all threads
+  for (const auto& [thread_id, dev_metrics_map] : *usage_metrics_map) {
+    bpt::ptree dev_array;
+    // iterate over all devices
+    for (const auto& [dev_id, dev_metrics] : dev_metrics_map) {
+      bpt::ptree dev;
+      dev.put("device_index", std::to_string(dev_id));
+      dev.put("bdf", dev_metrics.bdf);
+      dev.put("bos_peak_count", std::to_string(dev_metrics.bo_peak_count));
+
+      // add global bos
+      dev.add_child("global_bos", get_bos_ptree(dev_metrics.global_bos_met));
+
+      // add hw ctx info
+      dev.add_child("hw_context", get_hw_ctx_ptree(dev_metrics.hw_ctx_vec));
+
+      dev_array.push_back(std::make_pair("device", dev));
+    }
+
+    thread_array.add_child("thread " + std::to_string(t_count), dev_array);
+    t_count++;
+  }
+
+  print_json(thread_array);
+}
+
+} // namespace
+
+namespace xrt_core::usage_metrics {
```

Comment:
Everything below can stay in unnamed namespace except for the entry function `get_usage_metrics_logger()` at line 514.